### PR TITLE
Phase change patch 5: Adding the actual phase change module/code. Examples and tests not included yet.

### PR DESCRIPTION
**Pull request by amneetb**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by amneetb** [comment]
@boyceg: All of the code is now in here in these 5 patches. We will add examples in patch 6 and tests in patch 7. Not much would change in first 6 patches. Patch 7 will grow when we add more tests for some new features related to the first five patches.

**Comment by Raamkrishnan** [comment]
Rebased against master. This branch is ready for review now.

**Comment by amneetb** [comment]
@drwells : Let's get this branch merged soon. Would you have time to  review this patch?

**Comment by drwells** [comment]
I (finally) have some time to work on these things this week and next week.

File: include/ibamr/AllenCahnHierarchyIntegrator.h
Comment (drwells) [inline include/ibamr/AllenCahnHierarchyIntegrator.h:323]: Could you use non-Greek letters here and below? That way we don't need the extra comment.

File: src/phase_change/PhaseChangeHierarchyIntegrator.cpp
Comment (drwells) [inline src/phase_change/PhaseChangeHierarchyIntegrator.cpp:1194]: We do this in a small number of places elsewhere in the library but we shouldn't - the `d_` is to indicate that something is a member variable. It shouldn't be in the input database.

File: src/phase_change/AllenCahnHierarchyIntegrator.cpp
Comment (drwells) [inline src/phase_change/AllenCahnHierarchyIntegrator.cpp:15]:
```suggestion
```

File: src/phase_change/CarmanKozenyDragForce.cpp
Comment (drwells) [inline src/phase_change/CarmanKozenyDragForce.cpp:364]:
```suggestion
    avoid_division_by_zero_factor = db->getDouble("avoid_division_by_zero_factor");
```

Here and elsewhere - we should use the same names for everything in the input database/class/restart database (with the exception of the `d_` prefix)

File: src/phase_change/EnthalpyHierarchyIntegrator.cpp
Comment (drwells) [inline src/phase_change/EnthalpyHierarchyIntegrator.cpp:115]: This would be clearer if you flipped these around and encoded the default values in `getFromInput`. The same goes for the other classes - the input database always happens first, then restart data.

File: include/ibamr/CarmanKozenyDragForce.h
Comment (drwells) [inline include/ibamr/CarmanKozenyDragForce.h:67]:
```suggestion
                          SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
```
input databases should not be optional

File: include/ibamr/EnthalpyHierarchyIntegrator.h
Comment (drwells) [inline include/ibamr/EnthalpyHierarchyIntegrator.h:232]: Could you use `heat_capacity` instead of `Cp` here and mention in a comment that it is for (IIRC) constant pressure?

File: include/ibamr/MarangoniSurfaceTensionForceFunction.h
Comment (drwells) [inline include/ibamr/MarangoniSurfaceTensionForceFunction.h:194]: Why was this reverted? It's been long enough that I don't remember why we set this to NaN but I think it was done deliberately.

File: include/ibamr/PhaseChangeDivUSourceFunction.h
Comment (drwells) [inline include/ibamr/PhaseChangeDivUSourceFunction.h:121]:
```suggestion
```
AFAICT this object doesn't save things to the restart database so it does not need a name.

File: include/ibamr/HeavisideForcingFunction.h
Comment (drwells) [inline include/ibamr/HeavisideForcingFunction.h:100]:
```suggestion
```
Same

File: include/ibamr/PhaseChangeDivUSourceFunction.h
Comment (drwells) [inline include/ibamr/PhaseChangeDivUSourceFunction.h:101]:
```suggestion
```
These stray brackets might confuse doxygen.

File: include/ibamr/PhaseChangeHierarchyIntegrator.h
Comment (drwells) [inline include/ibamr/PhaseChangeHierarchyIntegrator.h:460]: Like elsewhere - can you expand out `Cp`?

File: src/phase_change/AllenCahnHierarchyIntegrator.cpp
Comment (drwells) [inline src/phase_change/AllenCahnHierarchyIntegrator.cpp:468]:
```suggestion
```

**Review by drwells — CHANGES_REQUESTED** [review]

File: src/phase_change/AllenCahnHierarchyIntegrator.cpp
Comment (drwells) [inline src/phase_change/AllenCahnHierarchyIntegrator.cpp:200]: This seems wrong - these variables are registered but the indices will go out of scope and thus we can never use them. What's going on here?

File: src/phase_change/AllenCahnHierarchyIntegrator.cpp
Comment (drwells) [inline src/phase_change/AllenCahnHierarchyIntegrator.cpp:867]: Here it is - we should store those patch data indices instead of looking them up just to match what's done elsewhere in IBAMR.

**Comment by drwells** [comment]
Neither this nor `phase_change_patch_7` contain tests for the following classes:
- AllenCahnHierarchyIntegrator
- PhaseChangeHierarchyIntegrator

can you add those here or there?

**Comment by Raamkrishnan** [comment]
@drwells Thanks for your suggestions. I will work on these suggestions.

File: include/ibamr/AllenCahnHierarchyIntegrator.h
Comment (Raamkrishnan) [reply include/ibamr/AllenCahnHierarchyIntegrator.h:341]: Thank you for your suggestion. I have now renamed all the Greek variables with their corresponding physical name. I use Greek names for temporary variables in a few places in ```.cpp``` files. For example:

https://github.com/IBAMR/IBAMR/blob/21f8114b1f29e644beceee4ce64c19dcc2da6bb9/src/phase_change/CarmanKozenyDragForce.cpp#L277

Can I do that?

File: src/phase_change/PhaseChangeHierarchyIntegrator.cpp
Comment (Raamkrishnan) [reply src/phase_change/PhaseChangeHierarchyIntegrator.cpp]: Thank you. I have corrected this everywhere.

File: src/phase_change/AllenCahnHierarchyIntegrator.cpp
Comment (Raamkrishnan) [reply src/phase_change/AllenCahnHierarchyIntegrator.cpp]: I have corrected this for all the variables. Please take a look.

File: include/ibamr/MarangoniSurfaceTensionForceFunction.h
Comment (Raamkrishnan) [reply include/ibamr/MarangoniSurfaceTensionForceFunction.h:194]: Even I forgot why did we revert to zero value. But me and Amneet discussed it over Zoom and decided to keep it as zero.

**Comment by Raamkrishnan** [comment]
> Neither this nor `phase_change_patch_7` contain tests for the following classes:
> 
> * AllenCahnHierarchyIntegrator
> * PhaseChangeHierarchyIntegrator
> 
> Can you add those here or there?

I can add a test for AllenCahnHierarchyIntegrator in ```phase_change_patch_7```. But I feel there is no need for the ```PhaseChangeHierarchyIntegrator``` class. Because it Is just a base class and does not do any computation.

**Comment by Raamkrishnan** [comment]
I have made changes to this patch based on the reviews. Please take a look. If everyone is happy, we can merge this. Then we have two more patches (examples and tests) to review and merge. That will conclude the phase change code implementation.

**Comment by drwells** [comment]
I could be convinced to skip tests for PhaseChangeIntegrator but we definitely need them for AllenCahnHierarchyIntegrator.

**Comment by Raamkrishnan** [comment]
> I could be convinced to skip tests for PhaseChangeIntegrator but we definitely need them for AllenCahnHierarchyIntegrator.

Sure. I will add the test for the AllenCahnHierarchyIntegrator class in ```phase_change_patch_7```. I already have two examples for the AllenCahnHierarchyIntegrator class in ```phase_change_patch_6```. I will use them for adding tests as well. This patch contains only the phase change implementation. Examples and tests for the phase change code are in ```phase_change_patch_6``` and ```phase_change_patch_7```. If everyone is happy with the implementation, we can merge this.

**Comment by drwells** [comment]
If you want to put the tests in another PR that's fine, but they need to exist before we merge this one.

File: include/ibamr/AllenCahnHierarchyIntegrator.h
Comment (drwells) [reply include/ibamr/AllenCahnHierarchyIntegrator.h:341]: Sure. The comment here still needs to be updated.

**Comment by drwells** [comment]
This looks good - the only substantial thing we need is tests. I don't see them on the phase change 7 branch. I'm happy to merge once there's a test for the AllenCahnHierarchyIntegrator somewhere in the pipeline.

**Comment by Raamkrishnan** [comment]
Sorry David for the delay. I will add the test today and push it.

**Comment by amneetb** [comment]
Some tests have failed in patch 7. See what is going on there.

--Amneet

**Comment by Raamkrishnan** [comment]
Hi David I have added a test for AllenCahn. But the result is not matching with the reference solution. I need to debug the test. I will try to fix this as soon as possible.

**Comment by Raamkrishnan** [comment]
@drwells I have added a test for AllenCahn's class in #1628. Please take a look. ```phase_change_patch_6 (#1627)``` and ```phase_change_patch_7 (#1628)``` are based on this patch. We can merge all these patches. I will now add a changelog for these patches and push it now.

**Comment by amneetb** [comment]
@drwells : Can you take a look at the three patches? If you are happy, we can merge and celebrate!

**Review by drwells — APPROVED** [review]

**Comment by drwells** [comment]
> @drwells : Can you take a look at the three patches? If you are happy, we can merge and celebrate!

Thanks!

I need to work on some more urgent things for paper review right now so I don't have time to look at the other patches today. I hope to get back to it soon.

**Pull request merged**
